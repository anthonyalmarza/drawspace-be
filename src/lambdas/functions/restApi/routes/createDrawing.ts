import { API, HandlerFunction } from 'lambda-api'
import cognitoAuth from 'lambdas/functions/restApi/middleware/cognitoAuth'
import createPublicDrawing from 'lambdas/util/accessPatterns/createPublicDrawing'
import createPrivateDrawing from 'lambdas/util/accessPatterns/createPrivateDrawing'
import validateRequest from 'lambdas/functions/restApi/middleware/validateRequest'

const validate = validateRequest(
    {
        title: { type: 'string', presence: true, length: { maximum: 255 } },
        thumbnailUrl: {
            type: 'string',
            presence: true,
            length: { maximum: 255 },
        },
        drawStepsUrl: {
            type: 'string',
            presence: true,
            length: { maximum: 255 },
        },
        startTime: { type: 'string', presence: true, length: { maximum: 27 } },
        endTime: { type: 'string', presence: true, length: { maximum: 27 } },
        width: { type: 'integer', presence: true },
        height: { type: 'integer', presence: true },
        resolution: { type: 'integer', presence: true },
        publish: { type: 'boolean', presence: true },
    },
    'body'
)

const handler: HandlerFunction = async (req) => {
    const { user, body } = req
    const { publish, ...rest } = body

    const params = { user: user.id, ...rest }
    let newDrawing
    if (publish) {
        newDrawing = await createPublicDrawing(params)
    } else {
        newDrawing = await createPrivateDrawing(params)
    }
    return newDrawing
}

export default (api: API): void => {
    api.post('/drawings', cognitoAuth, validate, handler)
}
