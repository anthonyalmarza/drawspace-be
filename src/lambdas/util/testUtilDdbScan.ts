import DynamoDB from 'aws-sdk/clients/dynamodb'

const { apiTableName } = require('lambdas/util/envVarsRestApi')
/* eslint-enable */
const DocumentClient = new DynamoDB.DocumentClient({
    endpoint: 'http://localhost:9001',
    accessKeyId: 'dynamo',
    secretAccessKey: 'devDummyKey',
    region: 'local',
})

export default () => DocumentClient.scan({ TableName: apiTableName }).promise()
