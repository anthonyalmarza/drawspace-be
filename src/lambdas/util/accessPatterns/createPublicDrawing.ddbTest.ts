import { advanceTo } from 'jest-date-mock'
import testUtilDdbScan from 'lambdas/util/testUtilDdbScan'

import createDrawing from 'lambdas/util/accessPatterns/createPublicDrawing'

advanceTo(new Date(2020, 5, 29, 0, 0, 0))

const mockUserId = '1234123412341234'

test('createPublicDrawing', async () => {
    const res = await createDrawing({
        id: 'mockId',
        title: 'mockTitle',
        thumbnailUrl: 'mockThumbnailUrl',
        drawStepsUrl: 'mockDrawStepsUrl',
        startTime: 'mockStartTime',
        endTime: 'mockEndTime',
        width: 1000,
        height: 1000,
        resolution: 2,
        user: mockUserId,
    })
    const scan = await testUtilDdbScan()
    expect(res).toMatchSnapshot()
    expect(scan).toMatchSnapshot()
})
