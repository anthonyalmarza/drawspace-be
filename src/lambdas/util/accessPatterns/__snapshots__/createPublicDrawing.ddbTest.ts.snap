// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createPublicDrawing 1`] = `
Object {
  "created": "2020-06-29T00:00:00.000Z",
  "drawStepsUrl": "mockDrawStepsUrl",
  "endTime": "mockEndTime",
  "entity": "drawing",
  "height": 1000,
  "id": "mockId",
  "modified": "2020-06-29T00:00:00.000Z",
  "published": "2020-06-29T00:00:00.000Z",
  "resolution": 2,
  "startTime": "mockStartTime",
  "thumbnailUrl": "mockThumbnailUrl",
  "title": "mockTitle",
  "user": "1234123412341234",
  "width": 1000,
}
`;

exports[`createPublicDrawing 2`] = `
Object {
  "Count": 1,
  "Items": Array [
    Object {
      "_ct": "2020-06-29T00:00:00.000Z",
      "_et": "drawing",
      "_md": "2020-06-29T00:00:00.000Z",
      "dsu": "mockDrawStepsUrl",
      "et": "mockEndTime",
      "gsi1pk": "gsi1pkDrawing#2020-06",
      "gsi1sk": "gsi1skDrawing#2020-06-29T00:00:00.000Z",
      "gsi2pk": "gsi2pkDrawing#1234123412341234",
      "gsi2sk": "gsi2skDrawing#public#2020-06-29T00:00:00.000Z",
      "he": 1000,
      "p": "2020-06-29T00:00:00.000Z",
      "pk": "drawing#mockId",
      "r": 2,
      "sk": "drawing#mockId",
      "st": "mockStartTime",
      "t": "mockTitle",
      "tu": "mockThumbnailUrl",
      "u": "1234123412341234",
      "wi": 1000,
    },
  ],
  "ScannedCount": 1,
}
`;
